libsolidity/ast/Types.cpp

```diff
@@ -700,6 +700,36 @@ TypePointer RationalNumberType::binaryOperatorResult(Token::Value _operator, Typ
 				value = rational(denominator, numerator);
 			break;
 		}
+		case Token::SHL:
+		{
+			using boost::multiprecision::pow;
+			if (isFractional() || other.isFractional())
+				return TypePointer();
+			else if (other.m_value < 0)
+				return TypePointer();
+			else if (other.m_value > numeric_limits<uint32_t>::max())
+				return TypePointer();
+			uint32_t exponent = other.m_value.numerator().convert_to<uint32_t>();
+			// Don't allow the result to take more than 1Mb of memory
+			if (((1024 * 1024 * 8) / m_value.numerator()) > exponent)
+				return TypePointer();
+			value = m_value.numerator() * pow(bigint(2), exponent);
+			break;
+		}
+		// NOTE: we're using >> (SAR) to denote (SHR)
```

Comment:
If you choose to implement SAR, please remove this comment (otherwise other people would follow the path I followed).